File: sprint5/By lessons/lesson12code1/lesson12code1.go
package main

type Node struct {
    left, right *Node
    size        int
}

func calculateSizes(root *Node) int {
    // Базовый случай рекурсии – это пустое дерево из нуля вершин.
    if root == nil {
        return 0
    }
    // Иначе суммируем размеры левого и правого поддерева
    // и прибавляем саму вершину root.
    root.size = 1 + calculateSizes(root.left) + calculateSizes(root.right)
    return root.size
}

File: sprint5/By lessons/lesson12code1/lesson12code1.cs
class Node {
    public Node left, right;
    public int size;
}

int CalculateSizes(Node root) {
    // Базовый случай рекурсии – это пустое дерево из нуля вершин.
    if (root == null) {
        return 0;
    }
    // Иначе суммируем размеры левого и правого поддерева
    // и прибавляем саму вершину root.
    root.size = 1 + CalculateSizes(root.left) + CalculateSizes(root.right);
    return root.size;
}
